### PR TITLE
Automatic allocation of external addresses for OVN load balancers and forwards

### LINKED ISSUE #11744: Allow automated allocation of external addresses (forward, load-balancers)
In a multitenant environment running with OVN, if one user in a project associates the external IP with the instance via forward - other users in other projects won't be able to track it. In this case one can never predict which IP they can pick for creating new forward.

### PR DESCRIPTION
**Pull request by markylaing**
Specification can be found here: https://discourse.ubuntu.com/t/automatic-listen-addresses-allocation-ovn-load-balancers-and-network-forwards/44441

Closes #11744

### CODE REVIEW

**Comment by tomponline** [comment]
@markylaing please can you move the contents of the spec doc to https://discourse.ubuntu.com/c/lxd/specifications/147

**Comment by markylaing** [comment]
> @markylaing please can you move the contents of the spec doc to https://discourse.ubuntu.com/c/lxd/specifications/147

Yep will do

**Comment by markylaing** [comment]
> @markylaing please can you move the contents of the spec doc to https://discourse.ubuntu.com/c/lxd/specifications/147

I've posted the spec publicly [here](https://discourse.ubuntu.com/t/automatic-listen-addresses-allocation-ovn-load-balancers-and-network-forwards/44441) and updated the PR description to link to the public one.

**Review by ru-fu — COMMENTED** [review]
This looks like it also needs an update to https://documentation.ubuntu.com/lxd/en/latest/howto/network_load_balancers/#create-a-network-load-balancer

(Note that this file is currently also updated in https://github.com/canonical/lxd/pull/13355 )

File: doc/api-extensions.md [hunk 1]
```diff
@@ -2407,3 +2407,8 @@ This extension adds a new image restriction, `requirements.nesting` which when `
 
 Adds the {config:option}`instance-miscellaneous:linux.kernel_modules.load` container configuration option. If the option is set to `ondemand`, the `finit_modules()` syscall is intercepted and a privileged user in the container's user namespace can load the Linux kernel modules specified in the
 allow list {config:option}`instance-miscellaneous:linux.kernel_modules`.
+
```
Comment (tomponline) [inline doc/api-extensions.md:2410]: When picking an API extension name, its worth looking at previous ones to try and keep to the naming pattern/theme.

The relevant ones I can find for this are:

`network_allocations`
`network_forward`
`network_load_balancer`

So we can say that this new API extension should start with `network_` at a minimum.

So how about: `network_allocate_external_ips`?

File: doc/api-extensions.md [hunk 2]
```diff
@@ -2407,3 +2407,8 @@ This extension adds a new image restriction, `requirements.nesting` which when `
...
+## `allocate_external_ips`
+
+Adds the ability to use an unspecified IPv4 or IPv6 address in the `listen_address` field of the request body for `POST /1.0/networks/{networkName}/load-balancers` and `POST /1.0/networks/{networkName}/forwards`.
```
Comment (tomponline) [inline doc/api-extensions.md:2413]: Can you add the example of an unspecified IP address in brackets for IPv4 and IPv6 here please.

File: lxd/network/driver_bridge.go [hunk 3]
```diff
@@ -3047,7 +3047,7 @@ func (n *bridge) getExternalSubnetInUse() ([]externalSubnetUsage, error) {
 }
 
 // ForwardCreate creates a network forward.
-func (n *bridge) ForwardCreate(forward api.NetworkForwardsPost, clientType request.ClientType) error {
+func (n *bridge) ForwardCreate(forward *api.NetworkForwardsPost, clientType request.ClientType) error {
```
Comment (tomponline) [inline lxd/network/driver_bridge.go:3050]: Can we instead of modifying the input param change this function signature to return a net.IP of the requested listen address? 

In general having functions modify their input arguments, but only in certain scenarios, can lead to confusion - whereas always returning the listen IP separately would be more predictable IMHO. What do you think?

File: lxd/network/network_utils.go [hunk 4]
```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
+// each randomly generated value. Each time validate returns an error, another IP address will be generated until the
+// context is cancelled.
+func randomAddressInSubnet(ctx context.Context, ipnet *net.IPNet, validate func(net.IP) error) (net.IP, error) {
```
Comment (tomponline) [inline lxd/network/network_utils.go:645]: Why pointer for net.IPNet here?

File: lxd/network/network_utils.go [hunk 5]
```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
...
+
+	// Can't get a random address from e.g. 192.0.2.1/32 as it's a single IP.
+	if quotient == bufferLen {
+		return nil, fmt.Errorf("Cannot find random address when subnet contains only one address")
+	}
+
+	for {
+		// Exit if the context is cancelled.
+		select {
+		case <-ctx.Done():
```
Comment (tomponline) [inline lxd/network/network_utils.go:672]: Can just check for ctx.Err() != nil here no need for switch.

File: lxd/network/network_utils.go [hunk 6]
```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
...
+
+	// The number of bytes that will be the same in the subnet and in the random address.
+	quotient := ones / 8
+
+	// The remaining bits that will be the same in the subnet and in the random address (after the equivalent bytes).
+	remainder := ones % 8
+
+	// Can't get a random address from e.g. 192.0.2.1/32 as it's a single IP.
+	if quotient == bufferLen {
+		return nil, fmt.Errorf("Cannot find random address when subnet contains only one address")
```
Comment (tomponline) [inline lxd/network/network_utils.go:666]: I would have expected this function to always return the same IP if a single /32 or /128 IP is passed in, rather than error.

Comment (tomponline) [reply lxd/network/network_utils.go:666]: I suppose this function should be keeping track of the offered IPs to the validate function and will not re-offer the same one repeatedly. And when there are no IPs left to offer fail out of the validate function hasn't returned nil by then.

Comment (tomponline) [reply lxd/network/network_utils.go:666]: Basically a single host value shouldnt be a special case :)

File: lxd/network/network_utils.go [hunk 7]
```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
...
+
+		var ip net.IP
+		if ipVersion == 4 {
+			ip = net.IPv4(r[0], r[1], r[2], r[3])
+		} else {
+			ip = r
+		}
+
+		// Skip the network address of the subnet itself.
+		if ip.Equal(ipnet.IP) {
```
Comment (tomponline) [inline lxd/network/network_utils.go:729]: What about the broadcast IP?

File: lxd/network/network_utils.go [hunk 8]
```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
...
+		// Read randomness into a byte slice of the correct length for the IP version.
+		r := make([]byte, bufferLen)
+		_, err := cryptorand.Read(r)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to automatically find an unused IPv%d address within %q: %w", ipVersion, ipnet.String(), err)
+		}
+
+		// Overwrite bytes/bits from the given subnet where the random address should be prefixed.
+		for i := 0; i <= quotient; i++ {
+			// For IPv4 addresses, the IP bytes SHOULD BE stored in the last four bytes of a 16 byte slice.
```
Comment (tomponline) [inline lxd/network/network_utils.go:686]: You should use ip.To4() to ensure you are using the 4 byte representation always.

> To4 converts the IPv4 address ip to a 4-byte representation.

Comment (tomponline) [reply lxd/network/network_utils.go:686]: either here or earlier in the function so we know what we are dealing with.

File: lxd/network/network_utils.go [hunk 9]
```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
...
+		// Exit if the context is cancelled.
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		// Read randomness into a byte slice of the correct length for the IP version.
+		r := make([]byte, bufferLen)
+		_, err := cryptorand.Read(r)
```
Comment (tomponline) [inline lxd/network/network_utils.go:679]: Could we simplify this process by taking the starting number of the subnet and then using https://pkg.go.dev/crypto/rand#Int to get a random number between 0 and the max size of the subnet and then adding it to the subnet?

File: lxd/network/network_utils.go (hunk 8 above)
Comment (tomponline) [reply lxd/network/network_utils.go:686]: This is what dnsmasq.DHCPAllAllocations does in the LXD code base btw.

Comment (tomponline) [reply lxd/network/network_utils.go:686]: This also allows storing the IPs as keys in map.

File: lxd/network/network_utils.go (hunk 9 above)
Comment (tomponline) [reply lxd/network/network_utils.go:679]: This is a bit like what getDHCPFreeIPv4() does in the dnsmasq dhpcalloc package, except its adding 1 each time.

File: lxd/network/driver_ovn.go [hunk 10]
```diff
@@ -214,6 +216,78 @@ func (n *ovn) projectRestrictedSubnets(p *api.Project, uplinkNetworkName string)
...
+				return err
+			}
+
+			return validator(ipnet)
+		})
+
+		subnetCtxCancel()
+
+		if err != nil {
+			if !errors.Is(err, context.DeadlineExceeded) {
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:278]: What if context.DeadlineExceeded comes from the parent context? We shouldnt go onto the next subnet in that case right?

Comment (tomponline) [reply lxd/network/driver_ovn.go:278]: I suggest checking ctx.Err() != nil here instead.

File: lxd/network/driver_ovn.go [hunk 11]
```diff
@@ -214,6 +216,78 @@ func (n *ovn) projectRestrictedSubnets(p *api.Project, uplinkNetworkName string)
...
+	if projectRestrictedSubnets == nil {
+		for _, uplinkRoute := range uplinkRoutes {
+			if (uplinkRoute.IP.To4() != nil && ipVersion == 4) || (uplinkRoute.IP.To16() != nil && ipVersion == 6) {
+				subnets = append(subnets, uplinkRoute)
+			}
+		}
+	}
+
+	if len(subnets) == 0 {
+		return nil, fmt.Errorf("No IPv%d routes are configured for this network", ipVersion)
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:245]: on this network's uplink.

Comment (tomponline) [reply lxd/network/driver_ovn.go:245]: Or, "... routes are available for this network"

File: lxc/network_forward.go [hunk 12]
```diff
@@ -308,7 +334,11 @@ func (c *cmdNetworkForwardCreate) Run(cmd *cobra.Command, args []string) error {
 	}
 
 	if !c.global.flagQuiet {
-		fmt.Printf(i18n.G("Network forward %s created")+"\n", forward.ListenAddress)
+		if c.flagAllocate != "" {
+			fmt.Printf(i18n.G("Network forward created") + "\n")
+		} else {
+			fmt.Printf(i18n.G("Network forward %s created")+"\n", forward.ListenAddress)
```
Comment (tomponline) [inline lxc/network_forward.go:340]: Lets put the listen address at the end to make it easier to parse, i.e `Network forward created: %s` and we can do this for all invocations.

File: lxd/network/driver_bridge.go (hunk 3 above)
Comment (markylaing) [reply lxd/network/driver_bridge.go:3050]: I was aiming to change the network driver interface as little as possible but I agree. Will change.

File: lxd/network/network_utils.go (hunk 4 above)
Comment (markylaing) [reply lxd/network/network_utils.go:645]: I don't have a good reason! A `net.IPNet` is mutable anyway as the underlying type of `net.IP` is a byte slice, but I agree we should pass by value if there is no good reason to pass by reference. Will change.

File: lxd/network/network_utils.go (hunk 5 above)
Comment (markylaing) [reply lxd/network/network_utils.go:672]: Ah of course :facepalm:

File: lxd/network/network_utils.go (hunk 6 above)
Comment (markylaing) [reply lxd/network/network_utils.go:666]: > I would have expected this function to always return the same IP if a single /32 or /128 IP is passed in, rather than error.
> Basically a single host value shouldnt be a special case - for the caller :)

In this case the only IP that can be returned would be the network address of the subnet itself, which is skipped below. This error essentially shortcuts the call so that we don't have to wait for the timeout.

> I suppose this function should be keeping track of the offered IPs to the validate function and will not re-offer the same one repeatedly. And when there are no IPs left to offer fail out of the validate function hasn't returned nil by then.

In my first iteration of this method I was figuring out the number of available IPs first, then picking a random number and mapping that back to the subnet. It turned out to be very complicated and I couldn't get it to pass my own unit tests :cry: 

Depending on the size of the subnet, keeping track of the IPs that have been attempted gets complicated quickly. E.g. if we have a map of attempted IPs, you could (very theoretically) have a map of size 2^128. In the end the simplest thing was to just generate random IPs until a timeout is reached.

File: lxd/network/network_utils.go (hunk 7 above)
Comment (markylaing) [reply lxd/network/network_utils.go:729]: Ok yep will add.

File: lxd/network/network_utils.go [hunk 13]
```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
```
Comment (roosterfish) [inline lxd/network/network_utils.go:642]: I was now thinking a bit on this function and maybe we could pick a slightly different route that is easier to follow?

If we get the first and last address from the given `ipnet`, this will give us two byte slices (regardless of IPv4 or IPv6). Those IPs (byte slices) could be transferred into an `uint64` so that we can use a random number generator in between those min and max limits. The resulting number could then be translated back to either IPv4 or IPv6 (I guess this is the tricky part). 

Another approach would be to use the `net/netip.Addr.Next()` functions until a free one is found. But this wouldn't use any randomness.

Comment (tomponline) [reply lxd/network/network_utils.go:642]: Thats why i was thinking to use https://pkg.go.dev/crypto/rand#Int

Comment (tomponline) [reply lxd/network/network_utils.go:642]: But yes, your suggestion is basically the same as https://github.com/canonical/lxd/pull/13379#discussion_r1579197346 but mine was to take off the prefix and then calculate a random host number

Comment (tomponline) [reply lxd/network/network_utils.go:642]: @roosterfish @markylaing what do you think of this:

```go
func randomIP(subnet net.IPNet) net.IP {
	// Take the starting IP of the subnet.
	startBig := big.NewInt(0)
	startBig.SetBytes(subnet.IP)

	// Calculate the host bits of the mask.
	ones, size := subnet.Mask.Size()
	hostBits := int64(size - ones)

	// Single host subnet.
	if hostBits <= 0 {
		return subnet.IP
	}

	// Calculate how many host IPs are available in the subnet.
	hostsBig := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(hostBits), nil)
	hostsBig.Sub(hostsBig, big.NewInt(1)) // Substract 1 so we don't go beyond the subnet.

	// Calculate the last IP in the subnet (optional, just for validation of logic).
	endBig := big.NewInt(0)
	endBig.Add(startBig, hostsBig)
	var endIP net.IP = endBig.Bytes()

	fmt.Printf("%v %v\n", subnet.IP.String(), endIP.String())

	// Calculate random IP in subnet.
	randBig, _ := rand.Int(rand.Reader, hostsBig)
	randBig.Add(startBig, randBig)
	var randIP net.IP = randBig.Bytes()

	return randIP
}
```

This seems to work well for both IPv4 and IPv6, including single host subnets.
What it doesn't cater for yet is avoiding returning the network or IPv4 broadcast addresses (there is no broadcast address in IPv6).

File: lxc/network_forward.go (hunk 12 above)
Comment (markylaing) [reply lxc/network_forward.go:340]: We might need to make a change to the client code to do this as the listen address is only returned in the `Location` header of the response, it's not returned by the client currently.

File: lxd/network/network_utils.go (hunk 13 above)
Comment (markylaing) [reply lxd/network/network_utils.go:642]: I'm working on this currently and this is where I'm at:
```golang
// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
// each randomly generated value. Each time validate returns an error, another IP address will be generated until all
// IPs have been exhausted or the context is cancelled.
func randomAddressInSubnet(ctx context.Context, ipnet net.IPNet, validate func(net.IP) error) (net.IP, error) {
	ones, _ := ipnet.Mask.Size()

	// networkAddress is the first address in the subnet.
	var networkAddress net.IP
	var subnetExponent int64
	if ipnet.IP.To4() != nil {
		subnetExponent = 32 - int64(ones)

		// Apply mask to get first address in subnet. Use To4 to ensure the IP is in the correct form for use with big.Int.
		networkAddress = ipnet.IP.To4().Mask(ipnet.Mask)
	} else {
		subnetExponent = 128 - int64(ones)
		networkAddress = ipnet.IP.To16().Mask(ipnet.Mask)
	}

	subnetSize := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(subnetExponent), nil)
	networkAddressInt := big.NewInt(0).SetBytes(networkAddress)

	// We want to exclude the network address and the broadcast address, so we'll exclude these from the range we generate
	// random numbers from.
	subnetSizeMinusTwo := big.NewInt(0).Sub(subnetSize, big.NewInt(2))

	// If the subnet is too small there are no addresses available. E.g. `192.0.2.0/31` contains only the network
	// address `192.0.2.0` and the broadcast address `192.0.2.1`.
	if subnetSizeMinusTwo.Cmp(big.NewInt(1)) <= 0 {
		return nil, noAvailableAddressErr{error: fmt.Errorf("No available addresses in subnet %q", ipnet.String())}
	}

	// Keep track of attempted IPs so that the validate function isn't repeatedly called in case it's expensive.
	// We don't need to worry about how large this map will grow. It is useful for small subnets where it's likely that
	// all IPs may be exhausted. The max size of a golang map will be the size of int, it is unfeasible for all of those
	// IP addresses to be in use (e.g. with `2001:db8::/32` the resource requirements for 2^32 IP addresses to be in use
	// precludes the possibility of this map growing too large).
	attempted := make(map[[16]byte]struct{})

	for {
		// Return on timeout or context cancellation.
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}

		// If we've attempted all possible addresses then return.
		if big.NewInt(int64(len(attempted))).Cmp(big.NewInt(0).Sub(subnetSizeMinusTwo, big.NewInt(1))) == 0 {
			return nil, noAvailableAddressErr{error: fmt.Errorf("No available addresses in subnet %q", ipnet.String())}
		}

		// Get a random number in range [0, subnetSizeMinusTwo).
		randBigInt, err := cryptorand.Int(cryptorand.Reader, subnetSizeMinusTwo)
		if err != nil {
			return nil, err
		}

		// Add the random number to the network address.
		randIPInt := big.NewInt(0).Add(networkAddressInt, randBigInt)

		// Add one to it to exclude the network address.
		randIP := net.IP(big.NewInt(0).Add(randIPInt, big.NewInt(1)).Bytes())
		randIP = randIP.To16()

		// Convert all IPs to a 16 byte representation so that we have a consistent map key we can use for both IPv4 and IPv6.
		var IPKey [16]byte
		copy(IPKey[:], randIP.To16())

		// Retry if we've already attempted this IP.
		_, ok := attempted[IPKey]
		if ok {
			continue
		}

		// Add the IP to the attempted map.
		attempted[IPKey] = struct{}{}

		// Call validate if set.
		if validate != nil {
			err = validate(randIP)
			if err != nil {
				continue
			}
		}

		return randIP, nil
	}
}
```

Comment (markylaing) [reply lxd/network/network_utils.go:642]: It's passing all my unit tests (I've added edge cases and small subnets in there as well)

Comment (roosterfish) [reply lxd/network/network_utils.go:642]: That looks quite easy to follow and doesn't even need the end IP. 

But do we really need to check if it's a multicast address? Wouldn't it be already a misconfiguration on the network side if the functions gets passed a `net.IPNet` in one of the assigned multicast ranges?

Comment (tomponline) [reply lxd/network/network_utils.go:642]: @roosterfish @markylaing updated my function now, seems to work well.

Comment (markylaing) [reply lxd/network/network_utils.go:653]: @tomponline thanks, it was definitely useful to talk through this with you both earlier! I've updated my version with the changes we discussed and updated to unit tests accordingly. If you can think of any more edge cases to add to the unit tests that would be very helpful. Thanks.

**Comment by markylaing** [comment]
@roosterfish @tomponline @ru-fu I've made all of the requested changes and have updated the documentation to include the `--allocate` flag and it's usage. I'm marking this as draft for now only because I haven't fully tested it yet. As soon as I've done that I'll unmark as draft and ping you again. In the meantime feel free to review. Thanks.

**Comment by markylaing** [comment]
@tomponline @ru-fu @roosterfish fixed a couple of bugs after manual testing. It's working nicely now. Ready for review when you have the time.

File: doc/api-extensions.md [hunk 14]
```diff
@@ -2407,3 +2407,10 @@ This extension adds a new image restriction, `requirements.nesting` which when `
 
 Adds the {config:option}`instance-miscellaneous:linux.kernel_modules.load` container configuration option. If the option is set to `ondemand`, the `finit_modules()` syscall is intercepted and a privileged user in the container's user namespace can load the Linux kernel modules specified in the
 allow list {config:option}`instance-miscellaneous:linux.kernel_modules`.
+
+## `network_allocate_external_ips`
+
+Adds the ability to use an unspecified IPv4 (`0.0.0.0`) or IPv6 (`::`) address in the `listen_address` field of the request body for `POST /1.0/networks/{networkName}/load-balancers` and `POST /1.0/networks/{networkName}/forwards`.
+If an unspecified IP address is used, supported drivers will allocate an available listen address automatically.
+Allocation of external IP addresses is currently supported by the OVN network driver.
+The OVN driver will allocate IP addresses from the subnets specified in the `ipv4.routes` and `ipv6.routes` configuration options.
```
Comment (ru-fu) [inline doc/api-extensions.md:2416]:
```suggestion
The OVN driver will allocate IP addresses from the subnets specified in the uplink network's `ipv4.routes` and `ipv6.routes` configuration options.
```

File: doc/howto/network_forwards.md [hunk 15]
```diff
@@ -34,11 +34,12 @@ In addition, proxy devices can be used to proxy traffic between different connec
...
 
 ```bash
-lxc network forward create <network_name> <listen_address> [configuration_options...]
+lxc network forward create <network_name> [<listen_address>] [--allocate=ipv{4,6}] [configuration_options...]
 ```
 
 Each forward is assigned to a network.
-It requires a single external listen address (see {ref}`network-forwards-listen-addresses` for more information about which addresses can be forwarded, depending on the network that you are using).
+A single external listen address might be required (see {ref}`network-forwards-listen-addresses` for more information about which addresses can be forwarded, depending on the network that you are using).
+If the network type supports IP allocation and a listen address is not given, the `--allocate` flag must be provided.
```
Comment (ru-fu) [inline doc/howto/network_forwards.md:42]:
```suggestion
Specify a single external listen address (see {ref}`network-forwards-listen-addresses` for more information about which addresses can be forwarded, depending on the network that you are using).
If the network type supports IP allocation, you don't need to specify a listen address.
If you leave it out, you must provide the `--allocate` flag.
```
Making the wording a bit more "active" (and hopefully clearer ;) ).

File: doc/howto/network_forwards.md [hunk 16]
```diff
@@ -50,7 +51,7 @@ Network forwards have the following properties:
 
 Property         | Type       | Required | Description
 :--              | :--        | :--      | :--
-`listen_address` | string     | yes      | IP address to listen on
+`listen_address` | string     | no       | IP address to listen on
```
Comment (ru-fu) [inline doc/howto/network_forwards.md:54]: Just marking this because this change will collide with https://github.com/canonical/lxd/pull/13355

File: doc/howto/network_load_balancers.md [hunk 17]
```diff
@@ -24,19 +24,20 @@ A load balancer is made up of:
...
+If a listen address is not given, the `--allocate` flag must be provided.
 
 ### Load balancer properties
 
 Network load balancers have the following properties:
 
 Property         | Type         | Required | Description
 :--              | :--          | :--      | :--
-`listen_address` | string       | yes      | IP address to listen on
+`listen_address` | string       | no       | IP address to listen on
```
Comment (ru-fu) [inline doc/howto/network_load_balancers.md:40]: Same comment about https://github.com/canonical/lxd/pull/13355

**Review by ru-fu — COMMENTED** [review]
Docs look good, just some small clarifications.

File: lxd/network/network_utils.go [hunk 18]
```diff
@@ -639,6 +639,111 @@ func randomSubnetV6() (string, error) {
...
+	if ipnet.IP.To4() != nil {
+		// Apply mask to get first address in subnet. Use To4 to ensure the IP is in the correct form for use with big.Int.
+		networkAddress = ipnet.IP.To4().Mask(ipnet.Mask)
+	} else {
+		networkAddress = ipnet.IP.To16().Mask(ipnet.Mask)
+	}
+
+	// If the subnet only has one address, or if the subnet is IPv4 and has two addresses return the network address.
+	if subnetExponent == 0 || (subnetExponent == 1 && ipnet.IP.To4() != nil) {
+		return networkAddress, nil
```
Comment (tomponline) [inline lxd/network/network_utils.go:671]: we still need to offer this address to `validate` right? Its just that if validate returns an error we don't try any more iterations in this case?

File: lxd/network/network_utils.go [hunk 19]
```diff
@@ -639,6 +639,111 @@ func randomSubnetV6() (string, error) {
...
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
+// each randomly generated value. Each time validate returns an error, another IP address will be generated until all
+// IPs have been exhausted or the context is cancelled.
+func randomAddressInSubnet(ctx context.Context, ipnet net.IPNet, validate func(net.IP) error) (net.IP, error) {
+	ones, size := ipnet.Mask.Size()
+	subnetExponent := size - ones
+
+	// networkAddress is the first address in the subnet.
+	var networkAddress net.IP
+	if ipnet.IP.To4() != nil {
```
Comment (tomponline) [inline lxd/network/network_utils.go:662]: Lets do this just once and store in a var, as its called a number of numbers in this function unnecessarily.

File: lxd/network/network_utils.go [hunk 20]
```diff
@@ -639,6 +639,111 @@ func randomSubnetV6() (string, error) {
...
+
+	// If the subnet is IPv6 and has two addresses, return the second one (no broadcast address for IPv6).
+	if subnetExponent == 1 {
+		return big.NewInt(0).Add(big.NewInt(0).SetBytes(networkAddress), big.NewInt(1)).Bytes(), nil
+	}
+
+	subnetSize := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(int64(subnetExponent)), nil)
+	networkAddressInt := big.NewInt(0).SetBytes(networkAddress)
+
+	// For Ipv4 we want to exclude the network address and the broadcast address, so we'll reduce the range we select
```
Comment (tomponline) [inline lxd/network/network_utils.go:682]: IPv4 not Ipv4 (like IPv6 in the above comment).

File: lxd/network/network_utils.go [hunk 21]
```diff
@@ -639,6 +639,111 @@ func randomSubnetV6() (string, error) {
...
+	}
+
+	// If the subnet only has one address, or if the subnet is IPv4 and has two addresses return the network address.
+	if subnetExponent == 0 || (subnetExponent == 1 && ipnet.IP.To4() != nil) {
+		return networkAddress, nil
+	}
+
+	// If the subnet is IPv6 and has two addresses, return the second one (no broadcast address for IPv6).
+	if subnetExponent == 1 {
+		return big.NewInt(0).Add(big.NewInt(0).SetBytes(networkAddress), big.NewInt(1)).Bytes(), nil
```
Comment (tomponline) [inline lxd/network/network_utils.go:676]: It shouldn't return it, but instead offer it to validate to check if its free right?

File: lxd/network/network_utils.go (hunk 20 above)
Comment (tomponline) [reply lxd/network/network_utils.go:682]: Can we do this calculation as part of the if block above on line 662, this is similar to my approach:

```go
	// Calculate how many host IPs are available in the subnet.
	usableHostsBig := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(int64(size-ones)), nil)

	// Standardise input.
	ip4 := subnet.IP.To4()
	if ip4 != nil {
		subnet.IP = ip4.Mask(subnet.Mask)
		usableHostsBig.Sub(usableHostsBig, big.NewInt(2)) // Remove network and broascast address.

	} else {
		subnet.IP = subnet.IP.To16().Mask(subnet.Mask)
		usableHostsBig.Sub(usableHostsBig, big.NewInt(1)) // Remove network address (ipv6 has no broadcast address).
	}
```

https://github.com/canonical/lxd/pull/13379/commits/33c3c73cf26cfcf263677242609d914601a845fc#r1580787239

As it keeps the version specific logic in one place.

File: lxd/network/network_utils.go [hunk 22]
```diff
@@ -639,6 +639,120 @@ func randomSubnetV6() (string, error) {
...
+		if ok {
+			continue
+		}
+
+		// Add the IP to the attempted map.
+		attempted[IPKey] = struct{}{}
+
+		// Call validate if set.
+		if validate != nil {
+			err = validate(randIP)
```
Comment (tomponline) [inline lxd/network/network_utils.go:746]: We should probably have validate return a `bool,error` such that if something goes wrong in the validation we dont take this incorrectly for being an IP not available.

File: lxd/network/driver_ovn.go [hunk 23]
```diff
@@ -214,6 +216,87 @@ func (n *ovn) projectRestrictedSubnets(p *api.Project, uplinkNetworkName string)
...
+
+	if projectRestrictedSubnets == nil {
+		for _, uplinkRoute := range uplinkRoutes {
+			if (uplinkRoute.IP.To4() != nil && ipVersion == 4) || (uplinkRoute.IP.To4() == nil && ipVersion == 6) {
+				subnets = append(subnets, uplinkRoute)
+			}
+		}
+	}
+
+	if len(subnets) == 0 {
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:244]: Feels like these 2 stanzas should be a case statement, it looks a bit odd calling `len` twice so close together.

File: lxd/network/driver_bridge.go [hunk 24]
```diff
@@ -3050,7 +3050,17 @@ func (n *bridge) getExternalSubnetInUse() ([]externalSubnetUsage, error) {
...
+		return nil, fmt.Errorf("Failed parsing address forward listen address %q: %w", forward.ListenAddress, err)
+	}
+
+	if listenAddressNet.IP.IsUnspecified() {
+		return nil, api.StatusErrorf(http.StatusNotImplemented, "Automatic listen address allocation not supported for drivers of type %q", n.netType)
+	}
+
+	err = n.state.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
 		// Check if there is an existing forward using the same listen address.
 		_, _, err := tx.GetNetworkForward(ctx, n.ID(), memberSpecific, forward.ListenAddress)
```
Comment (tomponline) [inline lxd/network/driver_bridge.go:3065]: Can we restructure this part to use the `externalSubnetsInUse` list (like you did for the `ovn` driver), just without adding support for automatic allocation, that way the structure of the code and the errors remains the same for each driver?

File: lxc/config/remote.go [hunk 25]
```diff
@@ -105,7 +131,7 @@ func (c *Config) GetInstanceServer(name string) (lxd.InstanceServer, error) {
 		return d, nil
 	}
 
-	// HTTPs
+	// HTTPS
```
Comment (tomponline) [inline lxc/config/remote.go:134]: Should be a separate commit really.

File: lxc/config/remote.go [hunk 26]
```diff
@@ -49,8 +50,39 @@ func (c *Config) ParseRemote(raw string) (remoteName string, resourceName string
...
+
+	// Check the remote is private.
+	if remote.Public || remote.Protocol == "simplestreams" {
+		return nil, fmt.Errorf("The remote isn't a private LXD server")
+	}
+
+	return remote, nil
+}
+
+// getPrivateRemoteByName returns the Remote with the given name.
```
Comment (tomponline) [inline lxc/config/remote.go:84]: doc block is wrong here?

File: lxc/network_forward.go [hunk 27]
```diff
@@ -277,38 +281,69 @@ func (c *cmdNetworkForwardCreate) Run(cmd *cobra.Command, args []string) error {
...
 		client = client.UseTarget(c.networkForward.flagTarget)
 	}
 
-	err = client.CreateNetworkForward(resource.name, forward)
+	err = client.CreateNetworkForward(networkName, forward)
 	if err != nil {
 		return err
 	}
 
+	parts := strings.Split(transporter.location, "/")
```
Comment (tomponline) [inline lxc/network_forward.go:342]: Need to improve the validation here before you extract the listenAddress to handle unexpected locations.

**Review by tomponline — DISMISSED** [review]

**Comment by tomponline** [comment]
Please can you rebase and then i can merge this, and you can address the comments in a follow up PR. Thanks!

**Comment by markylaing** [comment]
Rebased. I'll create an issue to follow up with remaining comments.

**Comment by markylaing** [comment]
@tomponline I've rebased this so you can merge when ready. Forgot to ping you, sorry.

**Review by tomponline — APPROVED** [review]

File: lxc/config/remote.go (hunk 25 above)
Comment (markylaing) [reply lxc/config/remote.go:134]: Won't address this as it has already been merged.
